### PR TITLE
Add missing password reset page to vue

### LINKED ISSUE #16540: Reset password from mail broken
Seems we broke this in #13929

Once you get a password reset link there are a few issues.

1. The CSRF token is not send
2. If you do send it you get back JSON

long story short. It seems we are missing some javascript here to properly handle the reset. And if the reset is done redirect to the normal login page.

### PR DESCRIPTION
**Pull request by juliusknorr**
Fixes https://github.com/nextcloud/server/issues/16540

I also removed the redirect we had, since we can now just show the prefilled login form on the same page.

### CODE REVIEW

File: core/src/components/login/UpdatePassword.vue [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+			<div id="submit-wrapper">
+				<input type="submit"
+					   id="submit"
+					   class="login primary"
+					   title=""
+					   :value="!loading ? t('core', 'Reset password') : t('core', 'Resetting password')" />
+				<div class="submit-icon"
+					 :class="{
+								'icon-loading-small': loading && invertedColors,
```
Comment (skjnldsv) [inline core/src/components/login/UpdatePassword.vue]: Indent issue

File: core/src/components/login/UpdatePassword.vue [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+				const data = {
+					password: this.user,
+					proceed: this.proceed
+				}
+
+				return Axios.post(this.resetPasswordTarget, data)
+					.then(resp => resp.data)
+					.then(data => {
+						let resetErrorMsg;
+						if (data && data.status === 'success'){this.message = 'send-success'
```
Comment (skjnldsv) [inline core/src/components/login/UpdatePassword.vue]:
```suggestion
						if (data && data.status === 'success'){
							this.message = 'send-success'
```

File: core/src/components/login/UpdatePassword.vue [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+							this.$emit('data:username', this.user)
+							this.$emit('done')
+						} else if (data && data.msg){
```
Comment (skjnldsv) [inline core/src/components/login/UpdatePassword.vue]:
```suggestion
						} else if (data && data.msg) {
```

File: core/src/components/login/UpdatePassword.vue [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+		},
+	}
+</script>
+
+<style scoped>
+	fieldset {
+		text-align: center;
+	}
+
+	input[ŧype=submit] {
```
Comment (skjnldsv) [inline core/src/components/login/UpdatePassword.vue]:
```suggestion
	input[type=submit] {
```

File: core/src/components/login/UpdatePassword.vue [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+  - along with this program. If not, see <http://www.gnu.org/licenses/>.
+  -
+  -->
+
+<template>
+	<form @submit.prevent="submit">
+		<fieldset>
+			<p>
+				<label for="password" class="infield">{{ t('core', 'New password') }}</label>
+				<input type="password" name="password" id="password" v-model="password" :placeholder="t('core', 'New password')" required />
```
Comment (skjnldsv) [inline core/src/components/login/UpdatePassword.vue]:
```suggestion
				<input type="password" name="password"
					id="password" v-model="password" required
					:placeholder="t('core', 'New password')" />
```

File: core/src/components/login/UpdatePassword.vue [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+			</p>
+
+			<div v-if="encrypted" class="update">
+				<p>
+					{{ t('core', 'Your files are encrypted. There will be no way to get your data back after your password is reset. If you are not sure what to do, please contact your administrator before you continue. Do you really want to continue?') }}
+				</p>
+				<input type="checkbox" class="checkbox" id="encrypted-continue" v-model="proceed" />
```
Comment (skjnldsv) [inline core/src/components/login/UpdatePassword.vue]:
```suggestion
				<input type="checkbox" class="checkbox"
					id="encrypted-continue" v-model="proceed" />
```

**Review by skjnldsv — DISMISSED** [review]
A few changes and typo :)
Great work ! :rocket:

**Comment by juliushaertl** [comment]
Thanks @skjnldsv ... all fixed.

File: core/src/components/login/UpdatePassword.vue [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+  - MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+  - GNU Affero General Public License for more details.
+  -
+  - You should have received a copy of the GNU Affero General Public License
+  - along with this program. If not, see <http://www.gnu.org/licenses/>.
+  -
+  -->
+
+<template>
+	<form @submit.prevent="submit">
```
Comment (skjnldsv) [inline core/src/components/login/UpdatePassword.vue]: shouldn't we stop as well?

File: core/src/components/login/UpdatePassword.vue [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+				this.loading = true
+				this.error = false
+				this.message = ''
+
+				const data = {
+					password: this.user,
+					proceed: this.proceed
+				}
+
+				return Axios.post(this.resetPasswordTarget, data)
```
Comment (skjnldsv) [inline core/src/components/login/UpdatePassword.vue]: I would prefer the `try...catch...finally` syntax if possible :wink: :see_no_evil:

File: core/src/components/login/UpdatePassword.vue (hunk 7 above)
Comment (juliushaertl) [reply core/src/components/login/UpdatePassword.vue]: Why? There is no other is involved in that page so propagation shouldn't do anytjing

Comment (skjnldsv) [reply core/src/components/login/UpdatePassword.vue]: submitting a form isn't supposed to reload the page on itself if no `action` is provided?

**Comment by rullzer** [comment]
Is this good to go now?

**Comment by juliushaertl** [comment]
ping @skjnldsv Should be all fixed as discussed yesterday

File: core/src/components/login/UpdatePassword.vue [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+				this.message = ''
+
+				try {
+					const { data } = await Axios.post(this.resetPasswordTarget, {
+						password: this.user,
+						proceed: this.proceed
+					})
+					if (data && data.status === 'success') {
+						this.message = 'send-success'
+						this.$emit('data:username', this.user)
```
Comment (skjnldsv) [inline core/src/components/login/UpdatePassword.vue]: Is this supposed to update the username variable?
Did you meant `update:username`?

File: core/src/views/Login.vue [hunk 10]
```diff
@@ -56,13 +56,20 @@
 								   @abort="resetPassword = false"/>
 				</div>
 			</div>
+			<div v-else-if="resetPasswordTarget !== ''">
+				<UpdatePassword :username.sync="user"
+							   :reset-password-target="resetPasswordTarget"
+							   :inverted-colors="invertedColors"
+							   @done="resetPasswordTarget = ''"/>
```
Comment (skjnldsv) [inline core/src/views/Login.vue]: I would prefer a dedicated method as it's cleaner :)

**Review by skjnldsv — APPROVED** [review]
Small nitpicks and comments.
Good otherwise! :)

File: core/src/views/Login.vue (hunk 10 above)
Comment (juliushaertl) [reply core/src/views/Login.vue]: Done.

File: core/src/components/login/UpdatePassword.vue (hunk 9 above)
Comment (juliushaertl) [reply core/src/components/login/UpdatePassword.vue]: Ah yes, thanks :)

**Review by rullzer — APPROVED** [review]

**Comment by ChristophWurst** [comment]
thank you!

**Pull request merged**